File: week-1/number-of-islands/number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        visited = [[False for i in grid[0]] for j in grid]
        count = 0
        
        for i in range(len(grid)):
            for j in range(len(grid[i])):
                if grid[i][j] == '1' and not visited[i][j]:
                    self.dfs(i, j, visited, grid)
                    count += 1
        
        return count
    
    def dfs(self, i, j, visited, grid):
        dx = [1, 0, -1, 0]
        dy = [0, 1, 0, -1]
        
        visited[i][j] = True
        q = [(i,j)]
        
        while q:
            cur_x, cur_y = q.pop()
            visited[cur_x][cur_y] = True
            
            for i in range(len(dx)):
                x = cur_x + dx[i]
                y = cur_y + dy[i]
                if x>=0 and y>=0 and x<len(grid) and y<len(grid[0]):
                    if grid[x][y] == '1' and not visited[x][y]:
                        q.append((x, y))
                        visited[x][y] = True
        

File: week-1/number-of-islands/number-of-islands.js
/**
 * @param {character[][]} grid
 * @return {number}
 */
var numIslands = function(grid) {
    let visited = new Array(grid.length).fill(false).map(i => Array(grid[0].length).fill(false));
    
    let count = 0;
    for (let i=0; i<grid.length; i++) {
        for (let j=0; j<grid[i].length; j++) {
            if (grid[i][j] === '1' && !visited[i][j]){
                dfs(i, j, grid, visited);
                count++;
            }
        }
    }
    
    return count;
};

let dfs = function (i, j, grid, visited) {
    if (i<0 || j<0 || i>grid.length-1 || j>grid[0].length-1 || grid[i][j] === '0' || visited[i][j])
        return;
    
    visited[i][j] = true;
    dfs(i+1, j, grid, visited);
    dfs(i-1, j, grid, visited);
    dfs(i, j-1, grid, visited);
    dfs(i, j+1, grid, visited);
};
